/*
 * GigaVoxels - GigaSpace
 *
 * Website: http://gigavoxels.inrialpes.fr/
 *
 * Contributors: GigaVoxels Team
 *
 * Copyright (C) 2007-2015 INRIA - LJK (CNRS - Grenoble University), All rights reserved.
 */

/** 
 * @version 1.0
 */

#include "CustomSectionEditor.h"

/******************************************************************************
 ******************************* INCLUDE SECTION ******************************
 ******************************************************************************/

// Qt
#include <QtCore/QUrl>
#include <QtGui/QFileDialog>
#include <QtGui/QMessageBox>
#include <QtGui/QVBoxLayout>
#include <QtGui/QToolBar>

// GvViewer
#include "Gvv3DWindow.h"
#include "GvvPipelineInterfaceViewer.h"
#include "GvvPluginManager.h"

#include "GvvApplication.h"
#include "GvvMainWindow.h"
#include "Gvv3DWindow.h"
#include "GvvPipelineInterfaceViewer.h"
#include "GvvPipelineInterface.h"

// Project
#include "SampleCore.h"

// STL
#include <iostream>

// System
#include <cassert>

// Qt
#include <QColorDialog>

/******************************************************************************
 ****************************** NAMESPACE SECTION *****************************
 ******************************************************************************/

// GvViewer
using namespace GvViewerCore;
using namespace GvViewerGui;

// STL
using namespace std;

/******************************************************************************
 ************************* DEFINE AND CONSTANT SECTION ************************
 ******************************************************************************/

/******************************************************************************
 ***************************** TYPE DEFINITION ********************************
 ******************************************************************************/

/******************************************************************************
 ***************************** METHOD DEFINITION ******************************
 ******************************************************************************/

/******************************************************************************
 * Default constructor
 ******************************************************************************/
CustomSectionEditor::CustomSectionEditor( QWidget *parent, Qt::WindowFlags flags )
:	GvvSectionEditor( parent, flags )
{
	setupUi( this );

	// Editor name
	setName( tr( "Rendering with distance fields" ) );
}

/******************************************************************************
 * Destructor
 ******************************************************************************/
CustomSectionEditor::~CustomSectionEditor()
{
}

/******************************************************************************
 * Populates this editor with the specified browsable
 *
 * @param pBrowsable specifies the browsable to be edited
 ******************************************************************************/
void CustomSectionEditor::populate( GvViewerCore::GvvBrowsable* pBrowsable )
{
	assert( pBrowsable != NULL );
	SampleCore* pipeline = dynamic_cast< SampleCore* >( pBrowsable );
	assert( pipeline != NULL );
	if ( pipeline != NULL )
	{
		// Parameters
		_opacitySpinBox->setValue( pipeline->getShapeOpacity() );
		_shaderMaterialPropertySpinBox->setValue( pipeline->getShaderMaterialProperty() );

		// Distance fields parameters
		_ceilingCheckBox->setChecked( pipeline->getCeilingState() );
		_floorCheckBox->setChecked( pipeline->getFloorState() );
		_columnsCheckBox->setChecked( pipeline->getColumnsState() );
		_monsterCheckBox->setChecked( pipeline->getMonsterState() );
		float x;
		float y;
		float z;
		pipeline->getEyePosition( x, y, z );
		_eyeXDoubleSpinBox->setValue( static_cast< double >( x ) );
		_eyeYDoubleSpinBox->setValue( static_cast< double >( y ) );
		_eyeZDoubleSpinBox->setValue( static_cast< double >( z ) );
		_mapScaleDoubleSpinBox->setValue( static_cast< double >( pipeline->getMapScale() ) );
		_bumpMappingCheckBox->setChecked( pipeline->useBumpMapping() );
		_ambientOcclusionCheckBox->setChecked( pipeline->useAmbientOcclusion() );
		_softShadowsCheckBox->setChecked( pipeline->useSoftShadows() );
	}
}

/******************************************************************************
 * Slot called when color tool button is released
 ******************************************************************************/
void CustomSectionEditor::on__colorToolButton_released()
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	const float3& shapeColor = sampleCore->getShapeColor();
	const float toIntColor = 255.f;
	const float toFloatColor = 1.f / 255.f;
	QColor color = QColorDialog::getColor( QColor( static_cast< int >( shapeColor.x * toIntColor ), static_cast< int >( shapeColor.y * toIntColor ), static_cast< int >( shapeColor.z * toIntColor ) ), this );
	if ( color.isValid() )
	{
		sampleCore->setShapeColor( make_float3( static_cast< float >( color.red() ) * toFloatColor, static_cast< float >( color.green() ) * toFloatColor, static_cast< float >( color.blue() ) * toFloatColor ) );
	}
}

/******************************************************************************
 * Slot called when opacity value has changed
 ******************************************************************************/
void CustomSectionEditor::on__opacitySpinBox_valueChanged( double pValue )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	sampleCore->setShapeOpacity( pValue );
}

/******************************************************************************
 * Slot called when shader material property (according to opacity) value has changed
 ******************************************************************************/
void CustomSectionEditor::on__shaderMaterialPropertySpinBox_valueChanged( double pValue )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	sampleCore->setShaderMaterialProperty( pValue );
}

/******************************************************************************
 * Slot called when ceiling distance field state has changed
 ******************************************************************************/
void CustomSectionEditor::on__ceilingCheckBox_stateChanged( int pValue )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	sampleCore->setCeilingState( pValue );
}

/******************************************************************************
 * Slot called when floor distance field state has changed
 ******************************************************************************/
void CustomSectionEditor::on__floorCheckBox_stateChanged( int pValue )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	sampleCore->setFloorState( pValue );
}

/******************************************************************************
 * Slot called when columns distance field state has changed
 ******************************************************************************/
void CustomSectionEditor::on__columnsCheckBox_stateChanged( int pValue )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	sampleCore->setColumnsState( pValue );
}

/******************************************************************************
 * Slot called when monster distance field state has changed
 ******************************************************************************/
void CustomSectionEditor::on__monsterCheckBox_stateChanged( int pValue )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	sampleCore->setMonsterState( pValue );
}

/******************************************************************************
 * Slot called when eye position has changed
 ******************************************************************************/
void CustomSectionEditor::on__eyeXDoubleSpinBox_valueChanged( double value )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	float x;
	float y;
	float z;
	sampleCore->getEyePosition( x, y, z );
	x = value;
	sampleCore->setEyePosition( x, y, z );
}

/******************************************************************************
 * Slot called when eye position has changed
 ******************************************************************************/
void CustomSectionEditor::on__eyeYDoubleSpinBox_valueChanged( double value )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	float x;
	float y;
	float z;
	sampleCore->getEyePosition( x, y, z );
	y = value;
	sampleCore->setEyePosition( x, y, z );
}

/******************************************************************************
 * Slot called when eye position has changed
 ******************************************************************************/
void CustomSectionEditor::on__eyeZDoubleSpinBox_valueChanged( double value )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	float x;
	float y;
	float z;
	sampleCore->getEyePosition( x, y, z );
	z = value;
	sampleCore->setEyePosition( x, y, z );
}

/******************************************************************************
 * Slot called when map scale has changed
 ******************************************************************************/
void CustomSectionEditor::on__mapScaleDoubleSpinBox_valueChanged( double value )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	sampleCore->setMapScale( value );
}

/******************************************************************************
 * Slot called when bump mapping state has changed
 ******************************************************************************/
void CustomSectionEditor::on__bumpMappingCheckBox_stateChanged( int pValue )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	sampleCore->setBumpMapping( pValue );
}

/******************************************************************************
 * Slot called when ambient occlusion state has changed
 ******************************************************************************/
void CustomSectionEditor::on__ambientOcclusionCheckBox_stateChanged( int pValue )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	sampleCore->setAmbientOcclusion( pValue );
}

/******************************************************************************
 * Slot called when soft shadows state has changed
 ******************************************************************************/
void CustomSectionEditor::on__softShadowsCheckBox_stateChanged( int pValue )
{
	GvvApplication& application = GvvApplication::get();
	GvvMainWindow* mainWindow = application.getMainWindow();
	Gvv3DWindow* window3D = mainWindow->get3DWindow();
	GvvPipelineInterfaceViewer* pipelineViewer = window3D->getPipelineViewer();
	GvViewerCore::GvvPipelineInterface* pipeline = pipelineViewer->editPipeline();

	SampleCore* sampleCore = dynamic_cast< SampleCore* >( pipeline );
	assert( sampleCore != NULL );

	sampleCore->setSoftShadows( pValue );
}
